packages/next/next-server/server/next-server.ts

```diff
@@ -963,11 +992,20 @@ export default class Server {
       return { html, pageData, sprRevalidate }
     })
 
-    // render fallback if cached data wasn't available
-    if (!isResSent(res) && !isDataReq && isDynamicRoute(pathname)) {
+    // render fallback if for a preview path or a non-seeded dynamic path
+    const isDynamicPathname = isDynamicRoute(pathname)
+    if (
+      !isResSent(res) &&
+      !isDataReq &&
+      ((isPreviewMode &&
+        // A header can opt into the blocking behavior.
+        req.headers['X-Prerender-Bypass-Mode'] !== 'Blocking') ||
+        isDynamicPathname)
+    ) {
       let html = ''
 
-      if (!this.renderOpts.dev) {
+      const isProduction = !this.renderOpts.dev
+      if (isProduction && (isDynamicPathname || !isPreviewMode)) {
```

Comment:
```suggestion
      if (isProduction && isDynamicPathname) {
```
We only generate fallbacks at build time for dynamic SSG pages currently unless this was changed somewhere I missed?

Reply:
This mimics the previous behavior.

Before, we'd unconditionally get the fallback in production mode:

```js
if (!this.renderOpts.dev) {
  html = await getFallback(pathname)
}
```

---

In the new behavior, `isProduction` must be joined with `isDynamicPathname` to ensure it exists (since we don't generate for others, as you said).

The `isPreviewMode` check will force this branch to behave as before, which was an "existing bug" so to speak.

If we'd like to fix this existing bug, I'd be happy to send a different PR.

Normally, it's impossible to request a fallback for a non-dynamic in non-preview mode because the cache entry would've resolved.

Reply:
Makes sense, as discussed in slack we can add an invariant here in a different PR 👍